atomix/cluster/src/main/java/io/atomix/raft/impl/ReconfigurationHelper.java

```diff
@@ -166,6 +170,170 @@ private void leaveInternal(final CompletableFuture<Void> future) {
+          newMembers);
+      final var newConfiguration =
+          new Configuration(
+              currentConfiguration.index() + 1,
+              raftContext.getTerm(),
+              Instant.now().toEpochMilli(),
+              newMembers,
+              Set.of(),
+              true);
+
+      raftContext.getCluster().configure(newConfiguration);
+    } else if (!(currentConfiguration.allMembers().containsAll(newMembers)
+        && newMembers.containsAll(currentConfiguration.allMembers()))) {
+      // This is not expected. When force configuration is retried, we expect that they are
+      // retried with the same state. If this is not the case, it is likely that there are two
+      // force configuration requested at the same time.
+      // Reject the request. There is possibly no way out to recover from this.
+      future.completeExceptionally(
+          new IllegalStateException(
+              String.format(
+                  "Expected to force configure with members '%s', but the member is already in force configuration with a different set of members '%s'",
+                  newMembers, currentConfiguration.allMembers())));
+      return;
+    }
+
+    sendForceConfigureRequestToAllMembers(future);
+  }
+
+  private void sendForceConfigureRequestToAllMembers(final CompletableFuture<Void> future) {
+    final Configuration configuration = raftContext.getCluster().getConfiguration();
+    final var otherMembers =
+        configuration.newMembers().stream()
+            .map(RaftMember::memberId)
+            .filter(m -> !m.equals(raftContext.getCluster().getLocalMember().memberId()))
+            .collect(Collectors.toSet());
+
+    final var quorum =
+        new ForceConfigureQuorum(
+            success -> {
+              if (Boolean.TRUE.equals(success)) {
+                future.complete(null);
+              } else {
+                future.completeExceptionally(
+                    new ProtocolException(
+                        "Failed to force configure because not all members acknowledged the request."));
+              }
+            },
+            otherMembers);
+
+    final ForceConfigureRequest request =
+        ForceConfigureRequest.builder()
+            .withTerm(configuration.term())
+            .withIndex(configuration.index())
+            .withTime(configuration.time())
+            .withNewMembers(configuration.newMembers())
+            .from(raftContext.getCluster().getLocalMember().memberId())
+            .build();
+
+    otherMembers.forEach(memberId -> sendForceConfigurationRequest(memberId, request, quorum));
+  }
+
+  private void sendForceConfigurationRequest(
+      final MemberId memberId,
+      final ForceConfigureRequest request,
+      final ForceConfigureQuorum quorum) {
+    logger.trace("Sending '{}' request to member '{}'", request, memberId);
+
+    raftContext
+        .getProtocol()
+        .forceConfigure(memberId, request)
+        .whenComplete(
```

Comment:
:clown_face: I always have a moment where I worry the future won't be on the right thread, but then I remember our protocol enforces it. Still, not sure if I like it :smile:

Reply:
Oops. Protocol do not enforce it. We have to explicitly pass the threadContext. :face_with_head_bandage:

Reply:
It doesn't? Oh, is it only on the RaftContext when we regisrer the handlers? So it's only on receive?

Reply:
Yes. :upside_down_face: